src/core/shape_inference/include/multiclass_nms_shape_inference.hpp

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <vector>
+
+#include <openvino/op/multiclass_nms.hpp>
+#include <openvino/op/util/multiclass_nms_base.hpp>
+
+namespace ov {
+namespace op {
+namespace util {
+
+template <class T>
+void shape_infer(const ov::op::util::MulticlassNmsBase* op,
+                 const std::vector<T>& input_shapes,
+                 std::vector<T>& output_shapes,
+                 bool static_output = false,
+                 bool ignore_bg_class = false) {
+    NODE_VALIDATION_CHECK(op, (input_shapes.size() == 2 || input_shapes.size() == 3)
+                    && output_shapes.size() == 3);
+
+    const auto& boxes_ps = input_shapes[0];
+    const auto& scores_ps = input_shapes[1];
+
+    const auto& nms_attrs = op->get_attrs();
+    const auto nms_top_k = nms_attrs.nms_top_k;
+    const auto keep_top_k = nms_attrs.keep_top_k;
+
+    // validate rank of each input
+    if (boxes_ps.rank().is_dynamic() || scores_ps.rank().is_dynamic()) {
+        return;
+    }
```

Comment:
Shouldn't this be `&&`? How do program reach L127/L146 otherwise? Is it even possible?

Reply:
no, we need both boxes_ps and scores_ps be static at the same time. So || is asked.
as to L127, we could remove this check, but it is also okay to keep it there to make them independent. How do you think?

Reply:
I see... I would remove it later as this is not possible to get into the branch anyways (reduction of code to understand and to execute). Actually if you think about it more, `ready_infer()` is already covered at the point of L43 when we apply logic that all of them need to be static. Isn't that true?